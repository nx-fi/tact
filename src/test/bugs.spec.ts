import { beginCell, toNano } from "@ton/core";
import { ContractSystem } from "@tact-lang/emulator";
import { __DANGER_resetNodeId } from "../grammar/ast";
import { SampleJetton } from "./bugs/output/bugs_SampleJetton";
import { JettonDefaultWallet } from "./bugs/output/bugs_JettonDefaultWallet";
import { Issue211 } from "./bugs/output/bugs_Issue211";

describe("bugs", () => {
    beforeEach(() => {
        __DANGER_resetNodeId();
    });
    it("should deploy sample jetton correctly", async () => {
        // Init
        const system = await ContractSystem.create();
        const treasure = system.treasure("treasure");
        const contract = system.open(
            await SampleJetton.fromInit(
                treasure.address,
                beginCell().endCell(),
                toNano("100"),
            ),
        );
        const target = system.open(
            await JettonDefaultWallet.fromInit(
                contract.address,
                treasure.address,
            ),
        );
        const tracker = system.track(target.address);
        await contract.send(
            treasure,
            { value: toNano("10") },
            {
                $$type: "Mint",
                receiver: treasure.address,
                amount: toNano("10"),
            },
        );
        await system.run();

        expect(contract.abi.errors!["31733"].message).toStrictEqual(
            "condition can`t be...",
        );

        expect(tracker.collect()).toMatchSnapshot();
    });
    it("should deploy issue211 correctly", async () => {
        // Init
        const system = await ContractSystem.create();
        const treasure = system.treasure("treasure");
        const contract = system.open(await Issue211.fromInit());
        const tracker = system.track(contract.address);
        await contract.send(treasure, { value: toNano("10") }, null);
        await system.run();

        expect(tracker.collect()).toMatchSnapshot();

        expect(await contract.getTest1()).toBe(0n);
        expect(await contract.getTest2()).toBe(0n);
        expect(await contract.getTest3()).toBe(6n);
        expect(await contract.getTest4()).toBe(24n);
        expect(await contract.getTest5()).toBe(97n);
        expect(await contract.getTest7()).toBe(42n);
        expect(await contract.getTest8()).toBe(5n);
        expect(await contract.getTest9()).toBe(5n);
    });
});
