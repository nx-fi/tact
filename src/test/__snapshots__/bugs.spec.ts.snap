// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bugs should deploy contract correctly 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{178D45190000000000000000502540BE4008001C6C7938BCE1EDDD4D7179F5626FBC23A92D0521900C93D0E579C0128C1BCFDB0016E3A425A4E75B646191AC9A34FE5D050BD101A5C490F87D01C66D885D09BC1082_}",
            "type": "cell",
          },
          "bounce": false,
          "from": "kQAONjycXnD27qa4vPqxN94R1JaCkMgGSehyvOAJRg3n7ROk",
          "to": "kQBW5ieoSEHXuEljVf_tAb_cPCvcurQZBoW1ez4bx54-Zadw",
          "type": "internal",
          "value": "9.959886",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 12277n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "type": "known",
              "value": {
                "$$type": "TokenExcesses",
                "queryId": 0n,
              },
            },
            "bounce": false,
            "from": "kQBW5ieoSEHXuEljVf_tAb_cPCvcurQZBoW1ez4bx54-Zadw",
            "to": "@treasure(treasure)",
            "type": "internal",
            "value": "9.916924834",
          },
        ],
      },
    ],
  },
]
`;
